### ISSUE TITLE
feature: prevent installation of rM2 incompatible apps on rM2

### ISSUE THREAD

**Issue by raisjn** [matteodelabre/toltec#129]
~~remux package should not recommend people disable xochitl, this caused someone on discord to break their rM2~~

it should be possible to prevent packages from being installed on rM2 if they are incompatible

**Comment by matteodelabre** [comment]
Since `how-to-enable` is based on the conflicts list of the systemd unit, can we consider this as a duplicate of #104?

**Comment by raisjn** [comment]
yes, you are right, the first part is duplicate (good catch!) - let's make the issue about the second part: specifying that apps are not installable on rM2

**Comment by Eeems** [comment]
So, how do we want to handle this? Separate repository for rm2?

**Comment by LinusCDE** [comment]
I propose having two kinda meta packages called `rm1-only` and `rm2-only` that have no files but a preinst check that fails when the model is wrong. The packages can also be made to conflict with each other.

Any package that does not support both devices can then just depend on those. In a case of wrong installation, the dependency would either fail in the preinst script (which should also be a fairly descriptive error to the user) or just conflict with the existing other `rmx-only` package.

To check the model, we could use either
- `/home/root/.config/remarkable/xochitl.conf`s `deviceid=` (which is the serial number that includes the model)
- `/sys/devices/soc0/machine`

to check the device is the correct one.

**Comment by Eeems** [comment]
I like that solution.

**Comment by LinusCDE** [comment]
I created a branch/pr to work on that solution: #136

**Comment by matteodelabre** [comment]
Thanks for your suggestion @LinusCDE! I see two drawbacks with your proposed approach:

1. If a package `X` needs to be compiled differently for RM1 and RM2 (e.g. it has a compile-time flag to flip to enable support for one or the other framebuffer), then if I understand correctly we’ll need to create two packages `X-rm1` and `X-rm2` that depend respectively on the `rm1-only` and `rm2-only` metapackages.
2. End users will see incompatible packages as installable in the Opkg package list and will not know that they are not compatible until they try to install them.

IMO, those limitations would be solved with a separate-repositories approach.

**Comment by LinusCDE** [comment]
@matteodelabre , those are some good ideas.

>     1. If a package `X` needs to be compiled differently for RM1 and RM2 (e.g. it has a compile-time flag to flip to enable support for one or the other framebuffer), then if I understand correctly we’ll need to create two packages `X-rm1` and `X-rm2` that depend respectively on the `rm1-only` and `rm2-only` metapackages.

That is certainly a drawback. But as it seems now, we are trying to try to make the new libs support both devices anyway.

I think that binaries that won't work without recompilation will be rare enough that they can solve that on their own to e.g. have a shell-script as binary that runs the correct binary or similar.

@Eeems how would this be done on QT? Can one easily swap the QT plugin for the rM1 to a potential one for the rM2? Even if this weren't possible, I'm sure others will peruse an approach them to have a single plugin work on both devices.

>     2. End users will see incompatible packages as installable in the Opkg package list and will not know that they are not compatible until they try to install them.

Most users will probably use either https://toltec-dev.org/stable/ or nao to look up packages. The website already having some pretty nice sorting and categorizing going on, we could surely add some badges to signify incompatibilities or just let users filter out incompatible ones. Something similar could be done for nao.

All in all, I think having multiple repositories is not worth the maintenance hassle for supporting the recompile approach of maybe one in 20 packages. Most will either use a library. For the 1/20 package I already proposed a solution above.

**Comment by matteodelabre** [comment]
I see your point. I propose we wait on this until we have at least a few apps supporting both devices, so that we can make an informed choice about the best suited strategy.

**Comment by LinusCDE** [comment]
Agreed, @matteodelabre :+1:

**Comment by raisjn** [comment]
>@Eeems how would this be done on QT? Can one easily swap the QT plugin for the rM1 to a potential one for the rM2? Even if this weren't possible, I'm sure others will peruse an approach them to have a single plugin work on both devices.

the rm2fb shim works with qt apps: oxide, calculator, draft have been tested.

i would like some ideas and help for how to package / move forward with this shim so everyone can use it easily. I can make an issue with open questions to discuss

**Comment by matteodelabre** [comment]
Users who try to install a rM2-only app on rM1 or vice versa will now see an error message (see #310).
